src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1383,12 +1392,74 @@ AddUnionAllSetOperationsToAttributeEquivalenceClass(AttributeEquivalenceClass **
 			continue;
 		}
 		int rtoffset = RangeTableOffsetCompat(root, appendRelInfo);
+		int childRelId = appendRelInfo->child_relid - rtoffset;
 
-		/* set the varno accordingly for this specific child */
-		varToBeAdded->varno = appendRelInfo->child_relid - rtoffset;
+		if (root->simple_rel_array_size <= childRelId)
+		{
+			/*
+			 * We always expect to find the childRel in the rte_array. Still,
+			 * being defensive is preferable.
```

Comment:
What is defensive about it?

Maybe: We prefer to return ever an Assert or error to be defensive.